### PR TITLE
Add `version_type` option to make version compatibly with semver

### PR DESCRIPTION
**Pull request by Apkawa**
## Description

I decided to use it in a rust project but cargo crashed when trying to use the prerelease versions.

I have added an optional setting that solves my problem for now.

### `version_type`

Choose version type

* `pep440` - default version provider.
    - prerelease - `1.0.1a0`
    - devrelease - `1.0.1dev0`
    - dev and pre - `1.0.1a0.dev0`
* `semver` - semver compatibly provider. Added "-" delimiter
    - prerelease - `1.0.1-a0`
    - devrelease - `1.0.1-dev0`
    - dev and pre - `1.0.1-a0-dev0`

TODO: renamed pre to long name, like alpha, beta, ..etc via additional option `semver_long`

```toml
[tool.commitizen]
version_type = "semver"
```

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Steps to Test This Pull Request
```shell
$ cz bump --check-consistency --no-verify --prerelease alpha --version-type=semver --dry-run
bump: version 0.1.0-a2 → 0.1.0-a3
tag to create: v0.1.0-a3

```

## Additional context

#150

### CODE REVIEW

**Comment by woile** [comment]
Hey this is amazing! Thanks for taking the time to do it.

We have to think of a different name instead of `version_provider`, as we are using that on `v3`, for the file providing the version, see [this](https://github.com/commitizen-tools/commitizen/blob/v3/docs/config.md#version-providers)

**Comment by Apkawa** [comment]
> Hey this is amazing! Thanks for taking the time to do it.
> 
> We have to think of a different name instead of `version_provider`, as we are using that on `v3`, for the file providing the version, see [this](https://github.com/commitizen-tools/commitizen/blob/v3/docs/config.md#version-providers)

hmm, maybe `version_format`?

**Comment by woile** [comment]
I was thinking of `version_type`, because version_format is the format used for the tag, e.g: if you want to have a tag `v$version`

File: commitizen/version_providers.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+        # Local version segment
+        if self.local is not None:
+            parts.append(f"+{self.local}")
+
+        return "".join(parts)
+
+
+providers = {
+    "pep": Version,
```
Comment (woile) [inline commitizen/version_providers.py:36]: Could we call this `pep440`? I'd rather give people an accurate keyword to search for

File: commitizen/version_providers.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from packaging.version import Version
+
+
+class SemVerVersion(Version):
```
Comment (woile) [inline commitizen/version_providers.py:4]: This is very clever. In my initial work I was thinking of creating a protocol for versions, like `VersionProtocol` and using [semver](https://pypi.org/project/semver/). 
1. Do you think re-implementing `__str__` could create any issues in the long term?
2. If we had to implement another version_type, let's say [calver](https://calver.org/) what should users implement?

File: tests/test_bump_find_version_provider.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import itertools
```
Comment (woile) [inline tests/test_bump_find_version_provider.py:1]: I think we should call this file `tests/test_bump_find_version_provider_semver.py` if we were to have more, I'd rather have one file per version

**Comment by woile** [comment]
Please also add the setting to [config.md](https://github.com/commitizen-tools/commitizen/blob/master/docs/config.md)

**Comment by woile** [comment]
@Lee-W wanna take a look? I left some minor comments, but in general it looks good. Do you have any thoughts or thing we should we aware of?

I'm thinking, once we merge this, we should rebase `v3`.

File: commitizen/version_providers.py (hunk 1 above)
Comment (Apkawa) [reply commitizen/version_providers.py]: done

**Comment by woile** [comment]
You can run `./scripts/format` to fix the linting errors.

File: commitizen/version_providers.py (hunk 2 above)
Comment (Apkawa) [reply commitizen/version_providers.py]: Look here please.
Only __str__ did not work out, I added to the protocol all the necessary methods that are used.

https://github.com/commitizen-tools/commitizen/pull/686/commits/f057e3f20cc026887a0d436922385a09f1af40ac#diff-ebe7b665130281170ae6aa39905fc0b3ba04a59b0d98787509c5aa3fce1b1b28R12

**Comment by Apkawa** [comment]
Oh, i checked only 3.9. Hmm

**Comment by Apkawa** [comment]
> Error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. (api.github.com:443)

Please restart checks

**Comment by Apkawa** [comment]
> Python package / python-check (3.7, windows-latest) (pull_request) 
image
hm, blinked test?

**Comment by Lee-W** [comment]
I'm on a trip these 2 weeks and probably won't be able to review it.

**Comment by AliSajid** [comment]
Bumping this since this is really important for me.

File: tests/commands/test_bump_command.py [hunk 4]
```diff
@@ -831,3 +831,168 @@ def test_bump_manual_version_disallows_prerelease_offset(mocker):
...
+        "--version-type",
+        "semver",
+    ]
+    mocker.patch.object(sys, "argv", testargs)
+    cli.main()
+
+    tag_exists = git.tag_exist("0.2.0-a0")
+    assert tag_exists is True
+
+    with open(tmp_version_file, "r") as f:
```
Comment (Lee-W) [inline tests/commands/test_bump_command.py:866]: nitpick: do you think we could try the following?

```python
for version_file in (tmp_version_file, tmp_commitizen_cfg_file):
    with open(version_file, "r") as f:
        assert "0.2.0-a0" in f.read()
```

Comment (Lee-W) [reply tests/commands/test_bump_command.py:866]: If you agree with this, maybe we could try this technique in other tests in this PR as well

File: tests/commands/test_bump_command.py [hunk 5]
```diff
@@ -831,3 +831,168 @@ def test_bump_manual_version_disallows_prerelease_offset(mocker):
...
+    tmp_version_file.write("0.1.0")
+    tmp_commitizen_cfg_file = tmp_commitizen_project.join("pyproject.toml")
+    tmp_version_file_string = str(tmp_version_file).replace("\\", "/")
+    tmp_commitizen_cfg_file.write(
+        f"{tmp_commitizen_cfg_file.read()}\n"
+        f'version_files = ["{tmp_version_file_string}"]\n'
+        f'version_type = "semver"'
+    )
+
+    create_file_and_commit("feat: new user interface")
```
Comment (Lee-W) [inline tests/commands/test_bump_command.py:901]: These setups are similar across different tests. Would it be possible for us to make them pytest fixtures?

File: tests/commands/test_changelog_command.py [hunk 6]
```diff
@@ -968,3 +968,60 @@ def test_empty_commit_list(mocker):
...
+    assert tag_exists is True
+
+    testargs = ["cz", "changelog", "0.3.0-a1", "--dry-run"]
+    mocker.patch.object(sys, "argv", testargs)
+    with pytest.raises(DryRunExit):
+        cli.main()
+
+    out, _ = capsys.readouterr()
+
+    assert out == "## 0.3.0-a1 (2022-02-13)\n\n"
```
Comment (Lee-W) [inline tests/commands/test_changelog_command.py:1027]: Should we unify the test way to use `pytest-regressions` as well?

File: tests/test_bump_find_version_type_semver.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    (("1.0.0a2", None, "beta", 1, None), "1.0.0-b1"),
+    (("1.0.0beta1", None, "rc", 0, None), "1.0.0-rc0"),
+    (("1.0.0rc1", None, "rc", 0, None), "1.0.0-rc2"),
+    (("1.0.0-a0", None, "rc", 0, None), "1.0.0-rc0"),
+    (("1.0.0-alpha1", None, "alpha", 0, None), "1.0.0-a2"),
+]
+
+
+@pytest.mark.parametrize(
+    "test_input,expected",
```
Comment (Lee-W) [inline tests/test_bump_find_version_type_semver.py:81]: nitpick: please add an additional space i.e., `test_input, expected`

**Review by Lee-W — APPROVED** [review]
@Apkawa This feature is amazing! Thanks for sending this awesome PR. I think we're almost ready to merge this one. I left a few minor comments. Let me know your thought.

File: tests/test_bump_find_version_type_semver.py (hunk 7 above)
Comment (Apkawa) [reply tests/test_bump_find_version_type_semver.py]: done

File: docs/config.md [hunk 8]
```diff
@@ -21,6 +21,7 @@
...
 | `major_version_zero`       | `bool` | `false`                     | When true, breaking changes on a `0.x` will remain as a `0.x` version. On `false`, a breaking change will bump a `0.x` version to `1.0`. [major-version-zero]                                                                                        |
 | `prerelease_offset`        | `int`  | `0`                         | In special cases it may be necessary that a prerelease cannot start with a 0, e.g. in an embedded project the individual characters are encoded in bytes. This can be done by specifying an offset from which to start counting. [prerelease-offset] |
+| `version_type`             | `str`  | `pep440`                    | Select a version type from the following options [`pep440`, `semver`]. Useful for non python projects. [See more][version_type]                                                                                                                      |
```
[the remark this answers is not in the document]
Comment (Apkawa) [reply docs/config.md]: done

File: tests/commands/test_bump_command.py (hunk 5 above)
Comment (Apkawa) [reply tests/commands/test_bump_command.py]: Ok. I wrote a simple fixture returning a function with parameters  https://github.com/commitizen-tools/commitizen/pull/686/commits/ea7ac936bc201e9090fa2d27cbae232e901fd4c1#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R30

**Comment by Apkawa** [comment]
hmm. I didn't touch dependencies. Do I need to merge the master onto a branch?

**Comment by Lee-W** [comment]
> hmm. CI failed. but i didn't touch dependencies. Do I need to merge the master onto a branch?

I just found out the root cause. It seems to relate to https://about.codecov.io/blog/message-regarding-the-pypi-package/. Could you please remove `codecov = "^2.0"` in `pyproject.toml` and try again? Thanks!

**Comment by Lee-W** [comment]
Other than that, everything looks good to me

**Comment by woile** [comment]
Could you point this to `v3`? `v3` removes codecov dep and the tests are working fine.

I'll be working on v3 this week, and I would like this to be part of it 💪🏻

**Comment by Apkawa** [comment]
Ok. I fixed CI.

**Comment by woile** [comment]
Could you point the PR to the branch `v3`? thanks

**Comment by Apkawa** [comment]
> Could you point the PR to the branch `v3`? thanks

Done. Failed not my test.

In #703 [test_changelog_incremental_with_merge_prerelease](https://github.com/commitizen-tools/commitizen/pull/703/files#diff-0f17f809d9a80303c2327a3d2830cecb7b76156ea47ad96b9dbbc877a08eb144R699)  need add `@pytest.mark.freeze_time("2023-04-16")` fixture.

**Comment by woile** [comment]
I'm asking for a rebase, not a squash. Otherwise the timeline will be dirty.
You can do something like:

```
git remote add upstream https://github.com/commitizen-tools/commitizen.git
git rebase upstream/v3
```
and then follow the instructions, fix the conflict and continue the rebase until it's done `git rebase --continue`

After that, you can do a force push and the PR will remain open

**Comment by Apkawa** [comment]
> I'm asking for a rebase, not a squash. Otherwise the timeline will be dirty. You can do something like:
> 
> ```
> git remote add upstream https://github.com/commitizen-tools/commitizen.git
> git rebase upstream/v3
> ```
> 
> and then follow the instructions, fix the conflict and continue the rebase until it's done `git rebase --continue`
> 
> After that, you can do a force push and the PR will remain open

done, please check

**Comment by woile** [comment]
Thank you so much! 🎉

**Pull request merged**